src/datarobotx/idp/custom_application_source.py

```diff
@@ -0,0 +1,320 @@
+#
+# Copyright 2024 DataRobot, Inc. and its affiliates.
+#
+# All rights reserved.
+#
+# DataRobot, Inc.
+#
+# This is proprietary source code of DataRobot, Inc. and its
+# affiliates.
+#
+# Released under the terms of DataRobot Tool and Utility Agreement.
+# https://www.datarobot.com/wp-content/uploads/2021/07/DataRobot-Tool-and-Utility-Agreement.pdf
+
+# mypy: disable-error-code="attr-defined"
+# pyright: reportPrivateImportUsage=false
+
+import contextlib
+import json
+import os
+from pathlib import Path
+from typing import Any, Dict, List, Optional, Tuple, Union
+
+from requests_toolbelt import MultipartEncoder
+
+import datarobot as dr
+from datarobot.models.runtime_parameters import RuntimeParameterValue
+from datarobot.utils import camelize
+
+from datarobotx.idp.common.hashing import get_hash
+
+
+def get_or_create_custom_application_source(endpoint: str, token: str, name: str) -> str:
+    """Get or create a custom application source with the given name.
+
+    Args:
```

Comment:
Small nit: in drx we followed [numpy docstring](https://numpydoc.readthedocs.io/en/latest/format.html) format. I think we're also following that here. Also in most of our helpers we did not explicitly write out endpoint and token in the docstring as arguments though I don't think we've ever explicitly said to exclude them.
https://github.com/datarobot-community/datarobotx-idp/blob/main/src/datarobotx/idp/common/handle_io.py#L101